dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/ScopedPointer.java

```diff
@@ -0,0 +1,66 @@
+package com.dat3m.dartagnan.program.memory;
+
+import com.dat3m.dartagnan.expression.Expression;
+import com.dat3m.dartagnan.expression.ExpressionKind;
+import com.dat3m.dartagnan.expression.ExpressionVisitor;
+import com.dat3m.dartagnan.expression.Type;
+import com.dat3m.dartagnan.expression.base.LeafExpressionBase;
+import com.dat3m.dartagnan.expression.type.IntegerType;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
+
+import java.util.Objects;
+
+public class ScopedPointer extends LeafExpressionBase<IntegerType> {
+
+    private final String id;
+    private final String scopeId;
+    private final Type innerType;
+    private final Expression address;
+
+    public ScopedPointer(String id, String scopeId, Type innerType, Expression address) {
+        super(TypeFactory.getInstance().getArchType());
+        this.id = id;
+        this.scopeId = scopeId;
+        this.innerType = innerType;
+        this.address = address;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public String getScopeId() {
+        return scopeId;
+    }
+
+    public Type getInnerType() {
+        return innerType;
+    }
+
+    public Expression getAddress() {
+        return address;
+    }
+
+    @Override
+    public ExpressionKind getKind() {
+        return ExpressionKind.Other.MEMORY_ADDR;
+    }
+
+    @Override
+    public <T> T accept(ExpressionVisitor<T> visitor) {
+        return getAddress().accept(visitor);
+    }
```

Comment:
You have to be careful here: an `ExpressionTransformer` that runs over the `ScopedPointer` would return the underlying `MemoryObject` and thus effectively strip away the scoping information.
In some cases this might be intended, but I think this can be a dangerous hidden interaction and at least deserves a comment.

Reply:
Yes, pointers can (and should) be improved/introduced, also for non-scoped memory. For the current use cases, it is intended to have the memory object here.